File: 프로그래머스/unrated/150368. 이모티콘 할인행사/이모티콘 할인행사.py
discount_rate = (10, 20, 30, 40)
ans = [0, 0]


def solution(users, emoticons):
    global ans
    n = len(emoticons)

    def dfs(discount_info: list):
        if len(discount_info) == n:
            tmp = [0, 0]
            for lower_discount_rate, upper_amount in users:
                amount = 0
                for i in range(n):
                    if discount_info[i] >= lower_discount_rate:
                        amount += int((100 - discount_info[i]) * emoticons[i] / 100)
                if amount >= upper_amount:
                    tmp[0] += 1
                else:
                    tmp[1] += amount
            global ans
            if tmp[0] > ans[0] or (tmp[0] == ans[0] and tmp[1] > ans[1]):
                ans = tmp
            return
        for rate in discount_rate:
            discount_info.append(rate)
            dfs(discount_info)
            discount_info.pop(-1)

    dfs([])
    return ans

File: 프로그래머스/unrated/150368. 이모티콘 할인행사/이모티콘 할인행사.java
import java.util.*;

class Solution {

	private static int n, emticonSize;
	private static int ansPrice = 0;
	private static int ansSubscribe = 0;
	private static final int[] discountAmount = {10, 20, 30, 40};// {0.1, 0.2, 0.3, 0.4};
	private static int[][] u;
	private static int[] e;

	public int[] solution(int[][] users, int[] emoticons) {
		n = users.length;
		emticonSize = emoticons.length;
		u = users;
		e = emoticons;

		dfs(new ArrayList<>());

		int[] ans = new int[2];
        ans[0] = ansSubscribe;
        ans[1] = ansPrice;
		return ans;
	}

	private void dfs(List<Integer> discountRates){
		if (discountRates.size() == emticonSize) {
			int tmpTotalPrice = 0, tmpSubNum = 0;

			for (int[] user: u) {
				int amount = 0;
				for (int i = 0; i < emticonSize; i++){
					if (discountRates.get(i) >= user[0]){
						amount += (int)((100 - discountRates.get(i)) * e[i] / 100);
					}
				}
				if (amount >= user[1]) {
					tmpSubNum++;
				} else {
					tmpTotalPrice += amount;
				}
			}

			if (ansSubscribe < tmpSubNum || (ansSubscribe == tmpSubNum && ansPrice < tmpTotalPrice)) {
				ansPrice = tmpTotalPrice;
				ansSubscribe = tmpSubNum;
			}
			return;
		}

		for (int da: discountAmount){
			discountRates.add(da);
			dfs(discountRates);
			discountRates.remove(discountRates.size() - 1);
		}
	}
}